### PR TITLE
VZ 3337: Istio Tracing

### PR DESCRIPTION
**Pull request by anders-swanson**
If Jaeger is enabled/installed, Istio will detect the Jaeger installation and configure itself to send Envoy traces to Jaeger (all apps in the Istio mesh).
- Envoy does not hot reload configuration. This means the pods would need to be restarted (either by the user, or by verrazzano) to start exporting traces.
- Istio sampling rate is 1% by default. We likely need to follow up another change to allow this to be configurable.

### CODE REVIEW

File: platform-operator/constants/constants.go [hunk 1]
```diff
@@ -166,3 +166,9 @@ const VerrazzanoIngressSecret = "verrazzano-tls" //nolint:gosec //#gosec G101
 
 // VerrazzanoLocalCABundleSecret is a secret containing the admin ca bundle
 const VerrazzanoLocalCABundleSecret = "verrazzano-local-ca-bundle" //nolint:gosec //#gosec G101
+
+//KubernetesAppLabel is a label key for kubernetes apps
+const KubernetesAppLabel = "app.kubernetes.io/component"
+
+//JaegerCollectorService is a label value for Jaeger collector
+const JaegerCollectorService = "service-collector"
```
Comment (kminder) [inline platform-operator/constants/constants.go:174]: Is this our label or a "standard" Jaeger label?  I'm trying to understand if it should be "name spaced".

File: platform-operator/controllers/verrazzano/controller.go [hunk 2]
```diff
@@ -1213,19 +1193,42 @@ func (r *Reconciler) watchPods(namespace string, name string, log vzlog.Verrazza
...
 
-	// Watch pods and trigger reconciles for Verrazzano resources when a pod is created
-	err := r.Controller.Watch(
-		&source.Kind{Type: &corev1.Pod{}},
-		mapFn,
-		predicateFunc)
-	if err != nil {
-		return err
+func (r *Reconciler) watchJaegerService(namespace string, name string, log vzlog.VerrazzanoLogger) error {
+	// Watch services and trigger reconciles for Verrazzano resources when the jaeger service is created
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/controller.go:1200]: Probably need to capitalize Jaeger in the comment.

File: platform-operator/controllers/verrazzano/controller.go [hunk 3]
```diff
@@ -1269,6 +1272,11 @@ func (r *Reconciler) initForVzResource(vz *installv1alpha1.Verrazzano, log vzlog
 		return newRequeueWithDelay(), err
 	}
 
+	if err := r.watchJaegerService(vz.Namespace, vz.Name, log); err != nil {
+		log.Errorf("Failed to set Service watch for Verrazzano CR %s: %v", vz.Name, err)
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/controller.go:1276]: Should the error message be about Jaeger vs Verrazzano?

File: platform-operator/constants/constants.go (hunk 1 above)
Comment (anders-swanson) [reply platform-operator/constants/constants.go:174]: It's a Jaeger label

Comment (kminder) [reply platform-operator/constants/constants.go:174]: ok

**Review by kminder — APPROVED** [review]
Approved

File: platform-operator/controllers/verrazzano/controller.go [hunk 4]
```diff
@@ -1185,20 +1173,12 @@ func (r *Reconciler) watchJobs(namespace string, name string, log vzlog.Verrazza
...
-					Name:      name,
-				}},
-			}
-		})
-
-	// Watch pod create
-	predicateFunc := predicate.Funcs{
-		CreateFunc: func(e event.CreateEvent) bool {
+	// Watch pods and trigger reconciles for Verrazzano resources when a pod is created
+	log.Debugf("Watching for pods to activate reconcile for Verrazzano CR %s/%s", namespace, name)
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/controller.go:1177]: Why are we doing this?  Seems like a very broad solution to force a full reconcile of all components.

File: platform-operator/controllers/verrazzano/controller.go [hunk 5]
```diff
@@ -1213,19 +1193,42 @@ func (r *Reconciler) watchPods(namespace string, name string, log vzlog.Verrazza
...
+}
 
-	// Watch pods and trigger reconciles for Verrazzano resources when a pod is created
-	err := r.Controller.Watch(
-		&source.Kind{Type: &corev1.Pod{}},
-		mapFn,
-		predicateFunc)
-	if err != nil {
-		return err
+func (r *Reconciler) watchJaegerService(namespace string, name string, log vzlog.VerrazzanoLogger) error {
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/controller.go:1199]: This component specific watch logic seems like it doesn't belong in the overall VZ CR reconcile.  Short term this is probably ok, but per component reconciles would be a better fit.

**Pull request merged**